File: tests/fixtures/feishu/capture_feishu.py
#!/usr/bin/env python3
"""
Feishu API fixture capture script.
Receives webhook events from Feishu bot and saves raw payloads as JSON files.

Usage:
    1. Run: python3 capture_feishu.py <port> <verification_token>
    2. Expose port via ngrok: ngrok http <port>
    3. Set the ngrok URL in Feishu Open Platform -> Event Subscription
    4. Send messages to the bot in Feishu to capture payloads

Output: Raw JSON files named by event_type and timestamp in this directory.
"""

import json
import os
import sys
import hashlib
import hmac
import time
from datetime import datetime
from http.server import HTTPServer, BaseHTTPRequestHandler

OUTPUT_DIR = os.path.dirname(os.path.abspath(__file__))
TOKEN = None

# Supported event types we care about for fixtures
EVENT_TYPES = {
    "im.message.receive_v1": "im-message",
    "im.message.message_read_v1": "im-message-read",
}


class FeishuHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        print(f"[{datetime.now().isoformat()}] {format % args}")

    def do_POST(self):
        if TOKEN is None:
            self.send_error_response(500, "No verification token set")
            return

        # Read raw body
        content_length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(content_length)

        # Validate signature
        # Feishu signs with: HMAC-SHA256 of verification_token + body
        sig = self.headers.get("X-Lark-Signature", "")
        expected = hmac.new(
            TOKEN.encode(), body, hashlib.sha256
        ).hexdigest()
        if not hmac.compare_digest(sig, expected):
            print(f"[WARN] Signature mismatch. Got {sig}")
            # Continue anyway for now — some events may not have signature
            # self.send_error_response(403, "Invalid signature")
            # return

        try:
            payload = json.loads(body)
        except json.JSONDecodeError as e:
            self.send_error_response(400, f"Invalid JSON: {e}")
            return

        event_type = payload.get("header", {}).get("event_type", "unknown")
        event_id = payload.get("header", {}).get("event_id", "no-event-id")

        # Save raw payload
        prefix = EVENT_TYPES.get(event_type, event_type.replace(".", "-"))
        ts = datetime.now().strftime("%Y%m%d-%H%M%S-%f")
        filename = f"{prefix}-{event_id}-{ts}.json"
        filepath = os.path.join(OUTPUT_DIR, filename)

        with open(filepath, "w", encoding="utf-8") as f:
            f.write(json.dumps(payload, ensure_ascii=False, indent=2))

        print(f"[SAVE] {filepath}  (event={event_type})")

        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(b'{"code":0,"msg":"ok"}')

    def send_error_response(self, code, msg):
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps({"code": code, "msg": msg}).encode())


def main():
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <port> <verification_token>")
        print(f"Example: {sys.argv[0]} 8080 your-verification-token")
        sys.exit(1)

    port = int(sys.argv[1])
    global TOKEN
    TOKEN = sys.argv[2]

    os.makedirs(OUTPUT_DIR, exist_ok=True)
    print(f"Listening on http://0.0.0.0:{port}")
    print(f"Output dir: {OUTPUT_DIR}")
    print(f"Verification token set: {TOKEN[:4]}...")
    print()
    print("Next steps:")
    print(f"  1. ngrok http {port}")
    print(f"  2. Copy the Forwarding URL (https://xxx.ngrok.io)")
    print(f"  3. Set it in Feishu Open Platform -> Event Subscription URL")
    print(f"  4. Send messages to your bot in Feishu")
    print()

    server = HTTPServer(("0.0.0.0", port), FeishuHandler)
    print(f"Server started. Press Ctrl+C to stop.")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\nStopped.")
        server.shutdown()


if __name__ == "__main__":
    main()

File: tests/fixtures/feishu/capture_feishu.js
#!/usr/bin/env node
/**
 * Feishu API fixture capture script.
 *
 * Uses Lark SDK's WSClient (WebSocket mode) to receive Feishu events.
 * No ngrok, no public URL, no verification token needed.
 *
 * Usage:
 *   node capture_feishu.js <app_id> <app_secret>
 *
 * Example:
 *   node capture_feishu.js cli_a964e565a9f8dcb3 PD2GqeCIgxWS1jVhbC1pHbgPwRNQFkVU
 *
 * Output: Raw JSON files named by event_type and timestamp in this directory.
 */

const Lark = require('@larksuiteoapi/node-sdk');
const fs = require('fs');
const path = require('path');

const OUTPUT_DIR = __dirname;

const APP_ID = process.argv[2];
const APP_SECRET = process.argv[3];

if (!APP_ID || !APP_SECRET) {
  console.error(`Usage: node ${path.basename(__filename)} <app_id> <app_secret>`);
  console.error(`Example: node ${path.basename(__filename)} cli_a964e565a9f8dcb3 PD2Gxxx`);
  process.exit(1);
}

// Event types we care about for fixtures
const EVENT_TYPES = [
  'im.message.receive_v1',
  'im.message.message_read_v1',
  'im.message.reaction.created_v1',
  'im.message.reaction.deleted_v1',
  'im.chat.member.bot.added_v1',
  'im.chat.member.bot.deleted_v1',
  'card.action.trigger',
  'drive.notice.comment_add_v1',
];

const client = new Lark.Client({
  appId: APP_ID,
  appSecret: APP_SECRET,
  appType: Lark.AppType.SelfBuild,
  domain: Lark.Domain.Feishu,
});

const wsClient = new Lark.WSClient({
  appId: APP_ID,
  appSecret: APP_SECRET,
  domain: Lark.Domain.Feishu,
  loggerLevel: Lark.LoggerLevel.info,
});

// Build event handlers
const handlers = {};
for (const eventType of EVENT_TYPES) {
  handlers[eventType] = async (data) => {
    const eventId = data.header?.event_id ?? `no-event-id`;
    const ts = new Date().toISOString().replace(/[:.]/g, '-');
    const safeType = eventType.replace(/\./g, '-');
    const filename = `${safeType}-${eventId}-${ts}.json`;
    const filepath = path.join(OUTPUT_DIR, filename);

    const payload = JSON.stringify(data, null, 2);
    fs.writeFileSync(filepath, payload, 'utf8');
    console.log(`[SAVE] ${filename}`);
    console.log(`  event_type: ${eventType}`);
    console.log(`  message_id: ${data.message?.message_id ?? 'N/A'}`);
    console.log(`  chat_id: ${data.message?.chat_id ?? 'N/A'}`);
    console.log(`  sender: ${data.message?.sender?.id ?? 'N/A'}`);
  };
}

const dispatcher = new Lark.EventDispatcher({
  encryptKey: '',
  verificationToken: '',
});
dispatcher.register(handlers);

console.log('='.repeat(60));
console.log('Feishu Fixture Capture Script (WebSocket Mode)');
console.log('='.repeat(60));
console.log(`App ID:  ${APP_ID}`);
console.log(`Output:   ${OUTPUT_DIR}`);
console.log();
console.log('Listening for events...');
console.log('Press Ctrl+C to stop.');
console.log();
console.log('Event types:');
for (const t of EVENT_TYPES) {
  console.log(`  - ${t}`);
}
console.log();
console.log('No ngrok needed! Lark SDK connects outbound via WebSocket.');
console.log('='.repeat(60));
console.log();

wsClient.start({ eventDispatcher: dispatcher }).catch((err) => {
  console.error('[ERROR] WSClient failed:', err);
  process.exit(1);
});

// Handle graceful shutdown
process.on('SIGINT', () => {
  console.log('\nStopping...');
  wsClient.close?.({ force: true });
  process.exit(0);
});
